packages/cache/src/internal/downloadUtils.ts

```diff
@@ -258,21 +260,41 @@ export async function downloadCacheStorageSDK(
         )
 
         downloadProgress.nextSegment(segmentSize)
-
-        const result = await client.downloadToBuffer(
-          segmentStart,
-          segmentSize,
-          {
+        const abortTimeInMs =
+          options.abortTimeInMs === undefined ? 3600000 : options.abortTimeInMs
+        const result = await promiseWithTimeout(
+          abortTimeInMs,
+          client.downloadToBuffer(segmentStart, segmentSize, {
+            abortSignal: controller.signal,
             concurrency: options.downloadConcurrency,
             onProgress: downloadProgress.onProgress()
-          }
+          })
         )
-
-        fs.writeFileSync(fd, result)
+        if (result === 'timeout') {
+          controller.abort()
+          throw new Error('Download aborted, segment download timed out.')
+        } else if (Buffer.isBuffer(result)) {
+          fs.writeFileSync(fd, result)
+        }
       }
     } finally {
       downloadProgress.stopDisplayTimer()
       fs.closeSync(fd)
     }
   }
 }
+
+const promiseWithTimeout = async (
+  timeoutMs: number,
+  promise: Promise<Buffer>
+): Promise<unknown> => {
+  let timeoutHandle: NodeJS.Timeout
+  const timeoutPromise = new Promise(resolve => {
+    timeoutHandle = setTimeout(() => resolve('timeout'), timeoutMs)
+  })
+
+  return Promise.race([promise, timeoutPromise]).then(result => {
+    clearTimeout(timeoutHandle)
```

Comment:
The promise timeout implementation looks very similar to what we were trying earlier but saw that promise still kept on awaiting till timeout was done. Is `clearTimeout` fixing that here?

Reply:
Yes, current implementation of [Timers promises API](https://nodejs.org/api/timers.html#timers-promises-api) only have an option to `setTimeout` but not clear it. So due to this, once the download was completed, there was no way to clear the timeout and the process would run for 1 hour straight. `clearTimeout` here will clear the timeout as soon as any of the promise is complete.

Reply:
That's interesting! I did not think not clearing timer will keep the thread locked

Reply:
Ah! it indeed is the case! https://nodejs.org/api/timers.html#class-timeout

> By default, when a timer is scheduled using either [setTimeout()](https://nodejs.org/api/timers.html#settimeoutcallback-delay-args) or [setInterval()](https://nodejs.org/api/timers.html#setintervalcallback-delay-args), the Node.js event loop will continue running as long as the timer is active.

Reply:
[Unref](https://nodejs.org/api/timers.html#timeoutunref) can be used to not block https://stackoverflow.com/a/43255345

Reply:
Yes! Btw, I am trying `AbortController.timeout(timeoutInMs)` right now and it is looking good for now. Working on error handling as the [error message](https://github.com/bbq-beets/test-cache-load/runs/7689143074?check_suite_focus=true#step:2:34) is not that good.